drivers/python/rethinkdb/net_asyncio.py

```diff
@@ -0,0 +1,236 @@
+    try:
+        yield
+    except asyncio.TimeoutError:
+        raise RqlTimeoutError
+
+
+# The asyncio implementation of the Cursor object:
+# The `new_response` Future notifies any waiting coroutines that the can attempt
+# to grab the next result.  In addition, the waiting coroutine will schedule a
+# timeout at the given deadline (if provided), at which point the future will be
+# errored.
+class AsyncioCursor(Cursor):
+    def __init__(self, *args, **kwargs):
+        Cursor.__init__(self, *args, **kwargs)
+        self.new_response = asyncio.Future()
+
+    def _extend(self, res):
+        Cursor._extend(self, res)
+        self.new_response.set_result(True)
+        self.new_response = asyncio.Future()
+
+    # Convenience function so users know when they've hit the end of the cursor
+    # without having to catch an exception
+    @asyncio.coroutine
+    def fetch_next(self, wait=True):
+        timeout = Cursor._wait_to_timeout(wait)
+        while len(self.items) == 0 and self.error is None:
+            self._maybe_fetch_batch()
+            with translate_timeout_errors():
+                yield from asyncio.wait_for(self.new_response, timeout)
+        # If there is a (non-empty) error to be received, we return True, so the
+        # user will receive it on the next `next` call.
+        return len(self.items) != 0 or not isinstance(self.error, RqlCursorEmpty)
+
+    def _empty_error(self):
+        # We do not have RqlCursorEmpty inherit from StopIteration as that interferes
+        # with Tornado's gen.coroutine and is the equivalent of gen.Return(None).
+        return RqlCursorEmpty(self.query.term)
+
+    @asyncio.coroutine
+    def _get_next(self, timeout):
+        while len(self.items) == 0:
+            self._maybe_fetch_batch()
+            if self.error is not None:
+                raise self.error
+            with translate_timeout_errors():
+                yield from asyncio.wait_for(self.new_response, timeout)
+        return convert_pseudo(self.items.pop(0), self.query)
+
+    def _maybe_fetch_batch(self):
+        if self.error is None and \
+           len(self.items) <= self.threshold and \
+           self.outstanding_requests == 0:
+            self.outstanding_requests += 1
+
+
+class ConnectionInstance(object):
+    _streamreader = None
+    _streamwriter = None
+
+    def __init__(self, parent, io_loop=None):
+        self._parent = parent
+        self._closing = False
+        self._user_queries = { }
+        self._cursor_cache = { }
+        self._ready = asyncio.Future()
+        self._io_loop = io_loop
+        if self._io_loop is None:
+            self._io_loop = asyncio.get_event_loop()
+
+        self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
+        self._socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+
+    @asyncio.coroutine
+    def connect(self, timeout):
+        try:
+            self._socket.connect((self._parent.host, self._parent.port))
```

Comment:
This call looks like it will block - checking out the `asyncio` API, I think you could combine this with `asyncio.open_connection` below.